Raising field to an integer power

TeranIvy opened the pull request:
This PR is for issue #90.

1. Branch is up-to-date with the master.
2. pep8 OK on modified .py files.
3. Cleaned up some pylint import issues (file 'dynamo0p3_builtins.py').
4. Added test for the newly introduced Built-in `inc_X_powint_n`.
5. Updated documentation (also updated all occurrences of `Built-in` to capital letter to be consistent throughout).
6. Tested the new Built-in in the LFRic code. All fine (build, run and results).

**Notes about changes**

I. There was no need to add support for a read-only integer scalar as all arguments can be read-only anyway. I added `gh_integer` to `VALID_BUILTIN_ARG_TYPES` in `dynamo0p3_builtins.py`.

II. Adding `gh_integer` resulted in `GenerationError` in `test_scalar_int_builtin_error(monkeypatch):` (file `dynamo0p3_builtins_test.py`). This happens because integer reductions are currently not supported in LFRic (file `dynamo0p3.py`) and it is a valid behaviour, as the code in the test attempts to store an inner product into an integer variable. I made two edits to address this problem.
a) In `dynamo0p3.py` I replaced `if scalar.type.lower() == "gh_integer":` with `if scalar.type.lower() == "gh_integer" and scalar.access.lower() == 'gh_sum':`. I suspect this addition is not really necessary, however I leave it to the reviewer to confirm this or not.
b) In `test_scalar_int_builtin_error(monkeypatch):` (file `dynamo0p3_builtins_test.py`) I replaced raising `ParseError` with raising `GenerationError`. `ParseError` was raised due to `gh_integer` not being allowed as an argument to a Built-in kernel and it is no longer valid. Test failure due to `GenerationError`, as said above, is a valid behaviour so I marked the test as `xfail` and stated the reason. I am not sure whether the modified `assert `statement is an overkill, though, because in other `xfail` tests full error strings are not used. 

Update on notes (09/11/2017): Marking `test_scalar_int_builtin_error(monkeypatch):` with `xfail` is not necessary so it is now removed.

**Notes about additional tests for `gh_integer`** (09/11/2017)

I went through `dynamo0p3_test.py` to check for `gh_integer` (and scalar in general) support in tests and I agree with @rupertford that a number of tests as well as the documentation will need to be updated. The question is whether to do it in this issue (PR) or open a new one (I am happy to do it here). My overview and proposition of changes is below.

1. Existing `gh_integer` support (no modification required)

- `test_int_scalar()`
- `test_two_int_scalars()`
- `test_two_scalars()`
- `test_dyninvoke_uniq_declns_intent_integer()`
- `test_stub_generate_with_scalars()`
- `test_scalars_only_invalid()`
- `test_kernel_stub_invalid_scalar_argument()`

2. Additional `gh_integer` support required (currently only tests for `gh_real`)

- `test_ad_scalar_type_too_few_args()`
- `test_ad_scalar_type_too_many_args()`
- `test_ad_scalar_type_no_write()`
- `test_ad_scalar_type_no_inc()`
- `test_no_vector_scalar()`
- `test_arg_descriptor_scalar_str()`

3. Possibly add `gh_integer` support (currently only tests for `gh_real`)

- `test_multiple_updated_scalar_args()` - The test replaces `(gh_real, gh_read)` with `(gh_real, gh_sum)`. I am not sure whether adding a similar replacement of `(gh_integer, gh_read)` with `(gh_real, gh_sum)` would add any value to the test, and `(gh_integer, gh_sum)` is not allowed in Dynamo 0.3 anyway.
- `test_arg_descriptor_repr()` - Should it pick up `DynArgDescriptor03(arg_type(gh_integer, gh_read))` in "testkern_qr_type" as well (it curently does that only for `DynArgDescriptor03(arg_type(gh_real, gh_read))`)?
- `test_stub_generate_with_scalar_sums()` - The test checks that the stub generator raises an exception when a kernel has a reduction. Does it make sense to add `gh_integer` case and, if so, what should the error message be (e.g. user-defined kernel is not allowed to have a reduction or integer reduction is not allowed in Dynamo 0.3)?

4. Modify error message

- `test_field_gh_sum_invalid()` 
- `test_operator_gh_sum_invalid()` 

   Both of these tests check that "reduction access `gh_sum` is only valid with a scalar argument". As Dynamo 0.3 API currently allows a reduction only to a real scalar, I think it would be useful to explicitly state this in the error message.
I.e. the error message should be "reduction access `gh_sum` is only valid with a real scalar argument" (modifying the related parse error in class `DynArgDescriptor03(Descriptor)`).

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2364,7 +2364,8 @@ def __init__(self, alg_invocation, idx):
                         arg_types=VALID_SCALAR_NAMES,
                         arg_accesses=VALID_REDUCTION_NAMES, unique=True):
-                    if scalar.type.lower() == "gh_integer":
+                    if scalar.type.lower() == "gh_integer" and \
```
I think this change is unnecessary since `args_filter` is being used to only return scalar arguments that are used in reductions (`arg_accesses=VALID_REDUCTION_NAMES`). Therefore if we've got an integer scalar at this point then we know that it's not valid.

TeranIvy replied:
I suspected this. Removed.

arporter commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -40,10 +40,10 @@
+from psyclone import psyGen
+from psyclone.psyGen import BuiltIn, NameSpaceFactory
+from psyclone.parse import ParseError
+from psyclone.dynamo0p3 import DynLoop, DynKernelArguments
```
Qualify imports - thanks.

TeranIvy replied:
Thanks.

arporter commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -595,13 +595,29 @@ def __str__(self):
+        integer_power = self._arguments.args[1].name
+        parent.add(AssignGen(parent, lhs=field_name,
+                             rhs=field_name + "**" + integer_power))
+
```
New class for built-in that raises a field to an integer power - good.

arporter commented on src/psyclone/dynamo0p3_builtins_mod.f90:
```diff
@@ -327,6 +327,18 @@ module dynamo0p3_builtins_mod
+   contains
+     procedure, nopass :: inc_X_powint_n_code
+  end type inc_X_powint_n
+
```
Meta-data for new built-in - good.

TeranIvy replied:
Thanks.

arporter commented on src/psyclone/tests/dynamo0p3_builtins_test.py:
```diff
@@ -1656,6 +1656,53 @@ def test_inc_X_powreal_a():  # pylint: disable=invalid-name
+                "loop\n"
+                "      !\n"
+                "      CALL f1_proxy%set_dirty()")
+
```
New test for new built-in - good.

arporter replied:
Please could you add a check (to this test) that the generated code compiles. Look for 'compile' elsewhere in this file or perhaps in `dynamo0p3_test.py`.

TeranIvy replied:
Added the check and happy to add checks for the other tests in the file as well as part of this issue. 
Just a warning that for me the compile fails with errors such as cited below.

   integer, parameter :: r_def     = real64
                                          1
   Error: Symbol 'real64' at (1) has no IMPLICIT type
/net/home/h06/ikavcic/PSyclone_work/PSyclone/src/psyclone/tests/test_files/dynamo0p3/infrastructure/constants_mod.f90:55.42:

   integer, parameter :: r_single  = real32
                                          1
   Error: Symbol 'real32' at (1) has no IMPLICIT type
/net/home/h06/ikavcic/PSyclone_work/PSyclone/src/psyclone/tests/test_files/dynamo0p3/infrastructure/constants_mod.f90:56.42:

   integer, parameter :: r_double  = real64
                                          1
   Error: Symbol 'real64' at (1) has no IMPLICIT type
/net/home/h06/ikavcic/PSyclone_work/PSyclone/src/psyclone/tests/test_files/dynamo0p3/infrastructure/constants_mod.f90:62.41:

   integer, parameter :: i_def     = int32

   Error: Symbol 'int32' at (1) has no IMPLICIT type
/net/home/h06/ikavcic/PSyclone_work/PSyclone/src/psyclone/tests/test_files/dynamo0p3/infrastructure/constants_mod.f90:80.17:

arporter replied:
I think it's best if we only touch the tests that you've already touched - small changesets are best!
If the compilation fails then this needs to be fixed as part of this ticket. Those symbols that the compiler is complaining about are from the iso_c_env module. Which compiler are you using?

TeranIvy replied:
Compilation report for "pytest --compile src/psyclone/tests/dynamo0p3_builtins_test.py" on master for different compilers we have on Desktop (branch behaves the same).

- GNU 5.3.0, 6.1.0 and 7.1.0: everything works.
- Intel 15.0.0, 16.0.1, 17.0.1: fails with the above error.
- Portland 15.7, 16.10: fails with the above error.

There should be a good support for iso_c_env module in all languages. I'll make some more checks and report back.

TeranIvy replied:
I just talked with Matthew, he thinks it's because compiler flags are set specifically for GNU. I'm checking that.

TeranIvy replied:
SOLVED: As the default is `gfortran`, all it needed was to explicitly specify `--f90="compiler name"`. Eg:

`py.test --compile --f90="pgfortran" src/psyclone/tests/dynamo0p3_builtins_test.py` for Portland

`py.test --compile --f90="ifort" src/psyclone/tests/dynamo0p3_builtins_test.py` for Intel

It is a bit slower than Gnu but it works :)

arporter replied:
Hurrah :-) Well done for finding the `--f90` flag and apologies for not mentioning it before.

arporter commented on src/psyclone/tests/dynamo0p3_builtins_test.py:
```diff
@@ -2570,6 +2617,9 @@ def test_multi_builtin_single_invoke():  # pylint: disable=invalid-name
+@pytest.mark.xfail(
+    reason="Reduction into an integer variable is not currently supported "
+    "by the LFRic infrastructure.")
 def test_scalar_int_builtin_error(monkeypatch):
```
I don't understand why this test is now marked as xfail. The purpose of the test is to check that PSyclone rejects an attempt to perform a reduction into an integer variable. This is still forbidden and therefore we still want to check that PSyclone rejects it with an appropriate error message.

TeranIvy replied:
I removed the xfail mark and tried to deal with the Generation Error as I saw in other tests. The test fails and I am not really sure what I did wrong. Help would be much appreciated, thank you!

TeranIvy replied:
SOLVED: As the default is gfortran, all it needed was to explicitly specify --f90="compiler name".

py.test --compile --f90="pgfortran" src/psyclone/tests/dynamo0p3_builtins_test.py for Portland

py.test --compile --f90="ifort" src/psyclone/tests/dynamo0p3_builtins_test.py for Intel

It is a bit slower than Gnu but it works :)

arporter commented on src/psyclone/tests/test_files/dynamo0p3/15.6.2_inc_X_powint_n_builtin.f90:
```diff
@@ -0,0 +1,48 @@
+
+  call invoke( inc_X_powint_n(f1, i_scalar) )
+
+end program single_invoke
```
New test file for new built-in - nice.

TeranIvy replied:
Thanks.

arporter commented on doc/dynamo0p3.rst:
```diff
@@ -1537,22 +1537,36 @@ where:
+where:
+
+* type(field_type), intent(inout) :: *field*
+* integer(i_def), intent(in) :: *iscalar*
```
Updated documentation - great!

TeranIvy replied:
Thanks.

arporter commented on doc/algorithm_layer.rst:
```diff
@@ -5,7 +5,7 @@ Algorithm layer
 
```
Documentation builds OK and looks good.

TeranIvy replied:
Thanks.

arporter commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -595,13 +595,29 @@ def __str__(self):
+        return "Built-in: raise a field to an integer power"
+
+    def gen_code(self, parent):
+        from psyclone.f2pygen import AssignGen
```
Please document the interface to this routine.

TeranIvy replied:
Documented interface to all "def gen_code" routines in the file.

arporter commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -595,13 +595,29 @@ def __str__(self):
     def gen_code(self, parent):
         from psyclone.f2pygen import AssignGen
         # In this case we're raising each element of a field to a
```
Please document the interface to this routine (sorry; I know you've barely touched it!)

TeranIvy replied:
Documented interface to all "def gen_code" routines in the file.

rupertford commented:
I'd like to take a look at this before we merge as I worry that we need to add a number of new tests and/or constraints to make sure gh_integer works, or can't be used in all possible situations. There is also the potential issue of updating documentation.

TeranIvy commented:
@arporter I made all the requested changes. I also removed the xfail mark from the `test_scalar_int_builtin_error(monkeypatch):` and tried to catch the Generation Error as I saw in other tests. The test fails nevertheless and I am not really sure what I am doing wrong, so help with this would be much appreciated.

@rupertford What tests did you have in mind?

arporter commented on src/psyclone/tests/dynamo0p3_builtins_test.py:
```diff
@@ -2583,12 +2638,15 @@ def test_scalar_int_builtin_error(monkeypatch):
+        # Attempt to generate the code
+        psy = PSyFactory("dynamo0.3",
+                         distributed_memory=dist_mem).create(invoke_info)
+        with pytest.raises(GenerationError) as excinfo:
```
I think the problem is that you've moved the check for the raising of `GenerationError` to be around the call to `psy.gen` whereas it still needs to be around the call to `PSyFactory(blah).create(blah)` - see how it was before.

TeranIvy replied:
Thanks! I thought that GenerationError was a property of psy.gen, hence moving the call. I tried several things and I'm afraid the test still doesn't work (list below).

1. Moved the check around PSyFactory as it was before and removed call to psy before the check.

        with pytest.raises(GenerationError) as excinfo:
            _ = PSyFactory("dynamo0.3",
                           distributed_memory=dist_mem).create(invoke_info)
        assert ("Integer reductions are not currently supported "
                "by the LFRic infrastructure. Error found in "
                "Kernel '" + test_builtin_name.lower() + "', argument 'isum'"
                in str(excinfo.value)

returns failure:

E               Failed: DID NOT RAISE <class 'psyclone.psyGen.GenerationError'>

The same happens if I move the call to psy within the check.

2. Put back the psy = PSyFactory(... call before the check. Now it fails with 

E                       GenerationError: "Generation Error: Integer reductions are not currently supported by the LFRic infrastructure. Error found in Kernel 'x_innerproduct_y', argument 'isum'"

3. Intentionally put the wrong info to check for ParseError (as it was before). It fails with GenerationError again.

arporter replied:
I've just had a look and realised that we fail to raise the error if distributed memory is not enabled!
We'll need to add a check for written, integer arguments earlier on - there must be a place where we validate the kernel arguments?

TeranIvy replied:
Got it now, thank you! I modified the test to call `PSyFactory` only  `if dist_mem` and I everything is now fine (I tested it for `True`). I am not sure what error to raise if `dist_mem` is not enabled, as the call to `PSyFactory` for this test only makes sense when it is enabled.

Regarding the checks for written, integer arguments you mentioned, I think a number of tests will need to be updated anyway (please see the updated description of this PR).

TeranIvy commented:
Back to @aporter for a second review. Please note that in reply to the request to add a test when kernel attempts to write to an integer scalar, as well as @rupertford's concerns about tests and documentation, I extended the test functionality since the last review.

**Notes about additional tests** (10/11/2017)

I added a number of new tests to check whether `gh_integer`scalar is behaving correctly in Dynamo 0.3 API. They were formed on the basis of existing tests which only checked for `gh_real` scalar behaviour before. The list of changes is below.

- `test_ad_scalar_type_too_few_args()` - renamed to `test_ad_real_scalar_type_too_few_args()` and created `test_ad_int_scalar_type_too_few_args()` 
- `test_ad_scalar_type_too_many_args()` - renamed to `test_ad_real_scalar_type_too_many_args()` and created `test_ad_int_scalar_type_too_many_args()`
- `test_ad_scalar_type_no_write()` - renamed to `test_ad_real_scalar_type_no_write()` and created `test_ad_int_scalar_type_no_write()`
- `test_ad_scalar_type_no_inc()` - renamed to `test_ad_real_scalar_type_no_inc()`  and created `test_ad_int_scalar_type_no_inc()` 
- `test_no_vector_scalar()` - renamed to `test_no_vector_real_scalar()` and created `test_no_vector_int_scalar()`
- `test_arg_descriptor_scalar_str()` - renamed to `test_arg_descriptor_real_scalar_str()` and created `test_arg_descriptor_int_scalar_str()`

Furthermore, as Dynamo 0.3 API currently allows a reduction only to a real scalar, I modified the related parse error message in class `DynArgDescriptor03(Descriptor)`) to explicitly state that `reduction access 'gh_sum' is only valid with a real scalar argument`. This led to a minor modification of these two tests:
- `test_field_gh_sum_invalid()` 
- `test_operator_gh_sum_invalid()` 

Finally, there are two tests which could also be extended to check for the allowed `gh_integer` behaviour. I did not modify them because I was not sure how much this would add to the checks already in place. My observations on them are below.

- `test_stub_generate_with_scalar_sums()` - The test checks that the stub generator raises an exception when a user-supplied kernel has a reduction. I think it would be useful to add `gh_integer` case as this is possible in a user-supplied kernel. However, I am not sure what should the error message be: a) user-defined kernel is not allowed to have a reduction or b) integer reduction is not allowed in Dynamo 0.3?

- `test_multiple_updated_scalar_args()` - The test replaces `(gh_real, gh_read)` with `(gh_real, gh_sum)`. As already said in the updated PR description, I am not sure whether adding a similar replacement of `(gh_integer, gh_read)` with `(gh_real, gh_sum)` would add any value to the test, and `(gh_integer, gh_sum)` is not allowed in Dynamo 0.3 anyway.

**Notes about documentation updates** (10/11/2017)

I edited the Dynamo 0.3 documentation to explicitly state that integer reductions are not currently allowed as they are not supported in the LFRic infrastructure. I did not remove the respective table entry on `GH_SUM` access (Built-ins only) for `GH_INTEGER` as I presume this is still planned for when the LFRic infrastructure allows this.
I also wrote a brief paragraph on scalars.

TeranIvy commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -213,6 +213,8 @@ def __str__(self):
 
     def gen_code(self, parent):
+        ''' Generates Dynamo0.3 API specific PSy code for a call to the
+            X_plus_Y Built-in. '''
```
Thanks for the review, @arporter. I am currently working on changes so I would like to check whether I got things right here.
- As a built-in call is essentially looping over DoFs so the parent type should be `psyclone.f2pygen.BaseGen`
- As for the `:param parent:` description, I noticed that in `dynamo0p3.py` the `psyclone.f2pygen.BaseGen` (in `gen_code`) docstrings were 
`        
:param parent: an f2pygen object that will be the parent of
 f2pygen objects created in this method
`
A call to built-in can be a part of a larger invoke object and it can also stand on its own. Does the `Node in f2pygen to which to add call` have similar meaning to the one above?

arporter replied:
At the level of f2pygen, we are dealing only with Fortran code - there's no relation back to PSyclone-type concepts such as looping over DoFs. `f2pygen.BaseGen` is the base class for all objects used to create Fortran and thus is the correct type for the parent.
Yes `Node in f2pygen to which to add call` is still correct - when generating Fortran the built-in will always be a child of something (e.g. SubroutineGen). Even when generating a kernel stub (which doesn't apply for built-ins), the routine will be a child of a module. It might be better to have `Node in f2pygen tree to which to ...` to emphasise that we are building a tree.

TeranIvy replied:
Thank you very much for the clarification. In the end I went with:

        :param parent: Node in f2pygen tree to which to add call
        :type parent: :py:class:`psyclone.f2pygen.BaseGen`

TeranIvy replied:
Thank you for the clarification. In the end I went with e.g.

        '''
        Generates Dynamo0.3 API specific PSy code for a call to the
        X_plus_Y Built-in.

        :param parent: Node in f2pygen tree to which to add call
        :type parent: :py:class:`psyclone.f2pygen.BaseGen`
        '''

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -63,7 +63,9 @@ Field Vector
-.. note:: To be written.
+In Dynamo 0.3 API a scalar is a single value variable that can be
+either real or integer. Real scalars are identified with ``GH_REAL``
+and integer scalars are identified with ``GH_INTEGER`` metadata.
```
Thanks! As advised, I didn't merge the master yet.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -510,7 +512,11 @@ For example:
-.. note:: In the Dynamo 0.3 API only :ref:`dynamo_built-ins` are permitted to write to scalar arguments (and hence perform reductions).
+.. note:: In the Dynamo 0.3 API only :ref:`dynamo_built-ins` are permitted
+          to write to scalar arguments (and hence perform reductions).
+          Furthermore, this permission is currently restricted for integer
```
Corrected.

TeranIvy replied:
Replaced, thanks.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -607,25 +613,29 @@ then the permitted access modes depend on the type of data it is and
+.. note:: As mentioned above, note that only Built-ins may modify
+          scalar arguments. *Due to the LFRic infrastructure currently
+          not supporting integer reductions there is a restriction
+          on using integer sums.*
```
Replaced, thanks for the suggestion.

TeranIvy replied:
Replaced.

TeranIvy commented:
Hi @aporter! I think I will have to merge the master to my branch as the new built-in test I added fails for some reason for PR build but works when just building on the branch.

aporter commented:
Hi all,

You’re sending this to the wrong aporter (Adam Porter).

Adam

> On Nov 14, 2017, at 12:25 PM, TeranIvy <notifications@github.com> wrote:
> 
> Hi @aporter! I think I will have to merge the master to my branch as the new built-in test I added fails for some reason for PR build but works when just building on the branch.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -154,9 +155,23 @@ def test_ad_scalar_type_too_few_args():  # pylint: disable=invalid-name
 
 
-def test_ad_scalar_type_too_many_args():  # pylint: disable=invalid-name
+def test_ad_int_scalar_type_too_few_args():  # pylint: disable=invalid-name
```
I guess "previous" means "test_ad_real_scalar_type_too_few_args". Merged using a loop.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -168,9 +183,23 @@ def test_ad_scalar_type_too_many_args():  # pylint: disable=invalid-name
 
 
-def test_ad_scalar_type_no_write():  # pylint: disable=invalid-name
+def test_ad_int_scalar_type_too_many_args():  # pylint: disable=invalid-name
```
I guess "previous" means "test_ad_real_scalar_type_too_many_args". Merged using a loop.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -182,9 +211,23 @@ def test_ad_scalar_type_no_write():  # pylint: disable=invalid-name
 
 
-def test_ad_scalar_type_no_inc():  # pylint: disable=invalid-name
+def test_ad_int_scalar_type_no_write():  # pylint: disable=invalid-name
```
I guess "previous" means "test_ad_real_scalar_type_no_write". Merged using a loop.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1371,6 +1428,20 @@ def test_no_vector_scalar():
         str(excinfo.value)
 
 
+def test_no_vector_int_scalar():
```
I guess "previous" means "test_no_vector_scalar". Merged using a loop to test for all scalars in `VALID_SCALAR_NAMES`.

TeranIvy commented on src/psyclone/tests/dynamo0p3_builtins_test.py:
```diff
@@ -2583,12 +2638,14 @@ def test_scalar_int_builtin_error(monkeypatch):
-        assert ("an argument to a built-in kernel must be one of ['gh_field', "
-                "'gh_real'] but kernel " + test_builtin_name.lower() + " has "
-                "an argument of type gh_integer" in str(excinfo))
+        if dist_mem:
```
I modified the line 551 in dynamo0p3.py to permit reductions into real scalars only:
        if self._type != "gh_real" and \
instead of
        if self._type not in VALID_SCALAR_NAMES and \

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -196,6 +239,20 @@ def test_ad_scalar_type_no_inc():  # pylint: disable=invalid-name
             "(['gh_sum']) but found 'gh_inc'" in str(excinfo.value))
 
 
+def test_ad_int_scalar_type_no_inc():  # pylint: disable=invalid-name
```
I guess "previous" means "test_ad_real_scalar_type_no_inc". Merged using a loop to test for all scalars in VALID_SCALAR_NAMES.

Instead of extending "test_ad_real_scalar_type_no_inc" to test that we reject GH_SUM for integer scalars I created a new test "test_ad_int_scalar_type_no_sum()" (different metadata descriptors).

As noted above, this test effectively checks for the same thing as test_scalar_int_builtin_error(monkeypatch)

TeranIvy commented:
Back to @arporter for another review. Please note potential issues with duplication of test functionality and slight decrease in coverage.

arporter commented:
Re-written/merged tests are really nice, documentation is up-to-date and docstrings are all lovely!
However, I will hold-off accepting this PR until Rupert's fix for the current problem on master (#103)  has been done. We will then need to tackle the job of bringing this branch up-to-date with master and fixing any conflicts.

TeranIvy commented:
Thanks for the update, @arporter. I was hoping to be able to merge this PR before the end of this week as I will be away until 6 December and cannot do anything about it. Can I at least update the branch with the latest master to facilitate the transition after Rupert fixes #103?

aporter commented:
I’m not @arporter.

Can you drop me from these messages please?

Thanks 

Adam

Sent from my iPhone

> On Nov 16, 2017, at 4:01 AM, TeranIvy <notifications@github.com> wrote:
> 
> Thanks for the update, @arporter. I was hoping to be able to merge this PR before the end of this week as I will be away until 6 December and cannot do anything about it. Can I at least update the branch with the latest master to facilitate the transition after Rupert fixes #103?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

arporter commented:
Hi @aporter - apologies for the spam. I think you have to unsubscribe yourself from this PR (top right); I can't see anyway for me to do it.

arporter commented:
Hi @TeranIvy - yes; I guess you can bring the branch up-to-date with master. Hopefully the fix that @rupertford is working on won't be too significant.

TeranIvy commented:
Merged the master and resolved conflicts in `dynamo0p3.py` (quite a lot of them, actually).

TeranIvy commented:
All tests pass after the merge of master. Back to @arporter for the final check and merge.
